plans/phase-1/api-design/multigres-operator-api-v1alpha1-design.md

```diff
@@ -0,0 +1,1121 @@
+2.  **Cluster-Level Default Template:** The corresponding template defined in `spec.templateDefaults` (e.g., `templateDefaults.coreTemplate` or `templateDefaults.cellTemplate`).
+3.  **Namespace-Level Default Template:** A template of the correct kind (e.g., `CoreTemplate`) named `default` in the same namespace as the cluster.
+4.  **Operator Hardcoded Defaults (Lowest):** A final fallback applied by the operator code (e.g., default resources, default replicas).
+
+#### Shard Override Chain
+*Applies to every Shard defined in `spec.databases[].tablegroups[].shards[]`.*
+
+1.  **Shard-Level Definition (Highest):** An inline `spec` or an explicit `shardTemplate` defined on the specific shard entry in the `MultigresCluster` YAML.
+2.  **Cluster-Level Default Template:** The `spec.templateDefaults.shardTemplate` field in the root `MultigresCluster` CR.
+3.  **Namespace-Level Default Template:** A `ShardTemplate` named `default` in the same namespace as the cluster.
+4.  **Operator Hardcoded Defaults (Lowest):** A final fallback applied by the operator.
+
+---
+
+### 2. Synchronous Validating Webhooks
+
+Webhooks are used *only* for fast, synchronous, and semantic validation to prevent invalid configurations from being accepted by the API server.
+
+#### `MultigresCluster`
+* **On `CREATE` and `UPDATE`:**
+    * **Template Existence:** Validates that all templates referenced in `spec.templateDefaults` or explicitly in components (Core, Cell, or Shard templates) exist *at the time of application*.
+    * **Component Spec Mutex:** Enforces that `managedSpec`, `external`, and `templateRef` are mutually exclusive for components like `GlobalTopoServer`.
+    * **Uniqueness:** Validates that all names are unique within their respective scopes:
+        * Cell names in `spec.cells`.
+        * Database names in `spec.databases`.
+        * TableGroup names within a Database.
+        * Shard names within a TableGroup.
+    * **Topology Integrity:** Verifies that if a Shard is pinned to a specific cell (via overrides), that cell exists in the `spec.cells` list.
+
+#### `CoreTemplate`, `CellTemplate`, `ShardTemplate`
+* **On `CREATE` and `UPDATE`:**
+    * **Schema Validation:** Ensures the template spec contains valid configuration for its specific type (e.g., a `CellTemplate` cannot contain `multiAdmin` config).
+
+---
+
+### 3. Asynchronous Controller and Finalizer Logic
+
+Asynchronous logic is used for operations that depend on external state or require blocking deletion, handled by controllers and finalizers.
+
+#### `MultigresCluster`
+* **Deletion Protection (Finalizer):**
+    1.  The `MultigresCluster` controller adds a finalizer (e.g., `multigres.com/cleanup`) to the CR upon creation.
+    2.  On deletion, the controller ensures all child resources (StatefulSets, PVCs, Services) are properly terminated before removing the finalizer.
+    3.  *Note:* Since databases are embedded in the cluster CR, deleting the cluster implies deleting all databases.
```

Comment:
We should probably mention how the deletion will delete most components, but will need to leave PVs detached and available for some time?